DataSourceVMwareGuestInfo.py

```diff
@@ -119,6 +120,10 @@ def get_data(self):
         # Get the vendor data.
         self.vendordata_raw = guestinfo('vendordata')
 
+        # Check to see if any of the guestinfo data should be removed.
+        if data_access_method == VMWARE_RPCTOOL:
+            clear_guestinfo_keys(self.metadata['cleanup-guestinfo'])
```

Comment:
Do we need to run `vmware-rpctool 'info-set guestinfo.userdata.cleared true'` after clear_guestinfo_keys() to explicitly indicate the userdata is readed and cleared? We need to tell CAPV this userdata is readed and should not be set again. Although CAPV can check if guestinfo.userdata equals to a single space to find out if userdata is cleared by cloud-init, but a dedicated key like 'guestinfo.userdata.cleared' seems more clear. Maybe vmware-rpctool will support removing a key in future version, then we don't need to set guestinfo.userdata to a single space instead of removing it.

Reply:
Hi @jessehu,

I don't think it's necessary to have an additional field to indicate the data has been cleared by the data-source. If CAPV or any other client really wants to know, all that's necessary is to inspect the data to see if it's empty or a single white-space character.

As for ensuring it's not set more than once, [CAPV assigns the cloud-init userdata during a clone operation](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/025bdc51fcc9d7260522b0927f2c3a73675109c0/pkg/services/govmomi/vcenter/clone.go#L45-L46), which occurs only once per machine.

For other clients I *could* see the argument that it's nice to know if the data was explicitly cleared. Instead of a separate key, however, I'm inclined to set the value to something like `---` or something to indicate the data has been read. That way:

* the value is till valid YAML
* it's evident the data has been read and cleared

Thoughts?

Reply:
Hi @jessehu,

I updated the clear value to `---`. A key's corresponding encoding key is set to the single whitespace character. This ensures the next time a cleared key is read, all that will be loaded is an empty YAML document.

Reply:
Thanks for the change. '---' is a clear indication for emtpy value 👍